core/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -219,6 +232,9 @@ public TopDocs call() throws Exception {
                     topDocsCollector = TopScoreDocCollector.create(numDocs, lastEmittedDoc);
                 }
                 collector = topDocsCollector;
+                if (doProfile) {
+                    collector = new InternalProfileCollector(collector, CollectorResult.REASON_SEARCH_TOP_HITS, Collections.emptyList());
```

Comment:
I don't mind giving it another try. Maybe we can get something that is a bit less magic than the previous version.